spyder/api/plugins.py

```diff
@@ -206,213 +328,380 @@ def check_compatibility(self):
+            self,
+            _("Undock"),
+            icon=ima.icon('undock'),
+            tip=_("Undock the pane"),
+            triggered=self._create_window,
+        )
+
+        self._close_plugin_action = create_action(
+            self,
+            _("Close"),
+            icon=ima.icon('close_pane'),
+            tip=_("Close the pane"),
+            triggered=self._plugin_closed,
+        )
+
+    # INFO: This is called from mainwindow.py
+    # This name does not convey that this can also (and should) be called
+    # from a non dockable widget. Non dockable widgets should be able to
+    # define a on_first_registration call
+    def _initialize_plugin_in_mainwindow_layout(self):
+        """
+        If this is the first time the plugin is shown, perform actions to
+        initialize plugin position in Spyder's window layout.
+
+        Use on_first_registration to define the actions to be run
+        by your plugin
+        """
+        if self.get_option('first_time', True):
+            try:
+                self.on_first_registration()
+            except NotImplementedError:
+                return
+            self.set_option('first_time', False)
+
+    # INFO: This is called from mainwindow.py
+    def _update_margins(self):
+        """Update plugin margins."""
+        layout = self.layout()
+        if self._default_margins is None:
+            self._default_margins = layout.getContentsMargins()
+
+        if self._config.get('main', 'use_custom_margin'):
+            margin = self._config.get('main', 'custom_margin')
+            layout.setContentsMargins(*[margin]*4)
+        else:
+            layout.setContentsMargins(*self._default_margins)
+
+    # INFO: This is called from mainwindow.py
+    def _create_dockwidget(self):
+        """Add to parent QMainWindow as a dock widget"""
+        # Creating dock widget
+        dock = SpyderDockWidget(self.get_plugin_title(), self.main)
+
+        # Set properties
+        dock.setObjectName(self.__class__.__name__ + '_dw')
+        dock.setAllowedAreas(self._ALLOWED_AREAS)
+        dock.setFeatures(self._FEATURES)
+        dock.setWidget(self)
+        self._update_margins()
+        dock.visibilityChanged.connect(self._visibility_changed)
+        dock.topLevelChanged.connect(self._on_top_level_changed)
+        dock.sig_plugin_closed.connect(self._plugin_closed)
+        self.dockwidget = dock
+
+        # NOTE: Don't use the default option of CONF.get to assign a
+        # None shortcut to plugins that don't have one. That will mess
+        # the creation of our Keyboard Shortcuts prefs page
+        # FIXME: All this logic should be simplified, and enforce the usage
+        # of CONF_DEFAULTS for users. And remove the "shortcut"
+        # attribute/special case. It is a little bit more work for the user
+        # but it makes things uniform and simplifies this code. We can add
+        # checks to ensure `switch to {}` matches the right name.
```

Comment:
This is not work for this PR.

Reply:
It is if I remove the `self.shortcut` logic plus the shortcut attribute